### DISCUSSION TITLE
exc_info simplifications

### DISCUSSION THREAD

**Discussion by iritkatriel** [faster-cpython/ideas#216] [General]
I'm trying to understand what it would take to make it possible to write (new) python code without knowing what (type, val, tb) is.  This is [my current draft](https://gist.github.com/iritkatriel/3927147548b10a7929cb0b680e3adc52).

I don't expect this to make python much faster, but it's the obvious followup to https://github.com/faster-cpython/ideas/issues/106.

**Comment by gvanrossum** [comment]
Sounds good, let's do it.

**Comment by iritkatriel** [comment]
I started the easy ones:

sys.exception(): https://bugs.python.org/issue46328
PyErr_GetActiveException/PyErr_SetActiveException: https://bugs.python.org/issue46343

**Comment by JelleZijlstra** [comment]
I opened https://bugs.python.org/issue46330 about the hard one, `__exit__`. I'd be interested in working with you to find the right approach to the migration. (For what it's worth, I'm skeptical that your suggested introspection approach will be sufficient, since there are likely to be non-introspectable `__exit__` methods out in the wild.)

**Reply by iritkatriel** [reply]
> (For what it's worth, I'm skeptical that your suggested introspection approach will be sufficient, since there are likely to be non-introspectable `__exit__` methods out in the wild.)

This is why I suggested to default to assuming that anything we don't understand has the legacy (triplet) signature. Only new code will use the new signature, so we don't need to be able to fully introspect all possible callable types that currently exist in the wild. As long as we document what we're doing this should be fine.  

I very much hope we don't need a migration-type approach that involves an opt-in API.

**Reply by JelleZijlstra** [reply]
But then wouldn't we have to maintain support for the old signature forever? I'm hoping that in 3.15 or so we'd be able to support only the new signature.

Another risk is that a user could apply a decorator to `__exit__` that makes the signature non-introspectable, and now suddenly it gets called with three arguments.

**Reply by iritkatriel** [reply]
> I'm hoping that in 3.15 or so we'd be able to support only the new signature.

I don't believe that's going to be possible.

**Reply by gvanrossum** [reply]
Do you think 3.15 is too soon or is there a fundamental reason?

**Reply by iritkatriel** [reply]
I think it's a big ask to make people change all the context mangers in their codebase, and given the stability slant of the new SC I don't think it would be a popular option now.   I'm not a stability fanatic, but I'm not sure whether I would vote for this one if I was on the SC.

**Reply by iritkatriel** [reply]
For sys.excepthook it could be easiest to add a new callback, say sys.exceptionhook(exc) which by default just calls sys.excepthook(type(exc), exc, exc.__traceback__) and make the interpreter call the new one. Then you can assign to either one of them and it would work.  (Then in python 4 we remove excepthook).

**Reply by iritkatriel** [reply]
On second though it would be better to have sys.excepthook call the new sys.exceptionhook, because there could be existing code out there that calls sys.excepthook directly and we need it to continue to work.

**Reply by gvanrossum** [reply]
Ah, indeed. When we add a PyFooBarEx() method because we want a new argument to PyFooBar(), we keep PyFooBar() around and make it call PyFooBarEx() with a suitable default for the new argument.

**Reply by pxeger** [reply]
But what happens when old code assigns to `sys.excepthook`? If the interpreter itself is still calling `sys.exceptionhook`, it won't attach the hook? Isn't this a catch-22?

(PS: if we're adding a new one with a new name, can we use `snake_case`?)

**Reply by iritkatriel** [reply]
In that case you want the user’s excepthook to take precedence so it’s fine. But if one module sets excepthook and another tries to override it but has been updated to use exceptionhook it can get interesting.

**Reply by iritkatriel** [reply]
Sorry - I misunderstood. The interpreter will keep calling excepthook.

**Reply by iritkatriel** [reply]
> But if one module sets excepthook and another tries to override it but has been updated to use exceptionhook it can get interesting.

so we need a function to register the new exceptionhook, which also resets excepthook to the default one.

**Reply by gvanrossum** [reply]
This exposes an inconsistency in API design for the sys module (not that we can do much about it now). Some APIs require you to call a setter function to change a callback (e.g. sys.setprofile()), others let you assign directly to a sys attribute. I like the setter function approach better. I don't recall why we didn't go for that with sys.excepthook (maybe because the corresponding state is consulted infrequently so we chose to inspect sys.__dict__ directly when it is needed rather than keep this in a C global somewhere).

**Reply by iritkatriel** [reply]
Possibly because sys.excepthook is expected to be called by user code and the profile function not?

**Reply by gvanrossum** [reply]
But if there was a setter, there could be a getter, and you could call it that way. The docs are not so clear.

I just hope that for the new API we'll use getter and setter functions.

**Reply by iritkatriel** [reply]
I wonder if we should perhaps defer the excepthook replacement for the future, because

1. It's not just sys.excepthook, there is also threading.excepthook.

2. the new APIs probably won't use used until all supported versions have the reduced exc_info (i.e., 3.11 is the earliest supported version) and in the meantime it's a messy API.  (excepthook + exceptionhook + unraisablehook + the threading equivalents).

**Reply by iritkatriel** [reply]
Also, since we concluded that the interpreter and the stdlib and the rest of the world will need to continue forever calling excepthook (which calls the new API), the benefit of this whole thing is questionable.

**Reply by gvanrossum** [reply]
Right, let's not rock this particular boat.

**Comment by markshannon** [comment]
Let's just leave `__exit__` as it is. 
Changing it is going to be painful and there is no advantage to be gained in terms of performance.
Many context managers don't care if an exception has been raised, so changing from `def __exit__(*args):` to `def __exit__(arg):` gains little in terms of usability.
